File: Sources/HerdrNotificationCore/NotificationEnvelope.swift
import CryptoKit
import Foundation

/// A decrypted Agent Notification payload: what a Host's notify hook tells
/// this device about an Agent Status transition (ADR 0008).
struct NotificationPayload: Sendable, Equatable {
    /// The herdr pane id the Agent lives in; drives the Attach deep link.
    let paneID: String
    /// The agent kind as herdr reports it (claude, codex, ...).
    let agentKind: String
    /// The new Agent Status. An open set like everywhere else on the herdr
    /// wire: unrecognized values pass through for forward compatibility.
    let status: AgentStatus
    /// When the transition happened.
    let timestamp: Date
}

/// The notification envelope v1: the encrypted payload carried from the
/// plugin's notify hook through the Push Relay and APNs to this app
/// (ADR 0008). Wire format, a compact JSON object:
///
///     {"v": 1, "kid": "...", "n": "...", "ct": "..."}
///
/// `ct` is the AES-256-GCM ciphertext (plus 16-byte tag) of a compact JSON
/// plaintext `{"pane", "kind", "status", "ts"}` under the per-host
/// Notification Key, with the version string bound as additional
/// authenticated data. The schema lives in `plugin/README.md`; the shared
/// test vectors in `plugin/test-vectors/notification-payload-v1.json` are
/// the single source of truth for this decoder and the plugin's encoder
/// (which proves the encrypt direction). Unknown fields at either layer are
/// ignored (additive v1 metadata); breaking changes bump the version, which
/// both implementations must honor together.
enum NotificationEnvelope {
    static let version = 1

    private static let keyBytes = 32
    private static let nonceBytes = 12
    private static let tagBytes = 16
    private static let keyIDBytes = 8

    /// Additional authenticated data binding the ciphertext to envelope v1,
    /// so a re-framed ciphertext under a different version fails to open.
    private static var additionalData: Data {
        Data("HERDR-NOTIFY:\(version)".utf8)
    }

    /// Derives the cleartext key id for a raw 32-byte Notification Key: the
    /// first 8 bytes of SHA-256 over the key, unpadded base64url. Both ends
    /// derive it, so it never needs to be stored or exchanged separately;
    /// the app uses it to select the right key (and thus Host) when several
    /// Hosts are registered.
    static func keyID(for key: Data) -> String {
        Data(SHA256.hash(data: key).prefix(keyIDBytes)).base64URLEncodedString()
    }

    /// Reads the cleartext key id off an envelope without decrypting it, so
    /// the right Notification Key can be selected when several Hosts are
    /// registered. Nil when the framing is too broken to carry one; version
    /// and ciphertext checks stay `decrypt`'s job.
    static func peekKeyID(in envelope: Data) -> String? {
        guard let wire = try? JSONDecoder().decode(WireEnvelope.self, from: envelope),
            let kid = wire.kid, !kid.isEmpty
        else { return nil }
        return kid
    }

    /// Decrypts an envelope's wire bytes with a Notification Key.
    ///
    /// Every way an envelope can be undecryptable — malformed framing, a
    /// future version, tampered or wrongly keyed ciphertext, a garbage
    /// plaintext — surfaces as a typed `NotificationEnvelopeError` so the
    /// service extension can fall back to a generic banner; nothing here
    /// crashes on hostile input.
    static func decrypt(
        _ envelope: Data, using key: Data
    ) throws(NotificationEnvelopeError) -> NotificationPayload {
        let wire: WireEnvelope
        do {
            wire = try JSONDecoder().decode(WireEnvelope.self, from: envelope)
        } catch {
            // Not a JSON object, or a field with the wrong JSON type
            // (including a non-numeric "v"): the framing is broken.
            throw .badEnvelope(reason: "envelope is not a well-formed JSON object")
        }
        guard let rawVersion = wire.v, let foundVersion = Int(exactly: rawVersion) else {
            throw .badEnvelope(reason: "v must be an integer")
        }
        guard foundVersion == version else {
            throw .unsupportedVersion(found: foundVersion)
        }
        guard let kid = wire.kid, !kid.isEmpty else {
            throw .badEnvelope(reason: "kid must be a non-empty string")
        }
        guard let nonceText = wire.n, let nonce = Data(base64URLEncoded: nonceText),
            nonce.count == nonceBytes
        else {
            throw .badEnvelope(reason: "n must be \(nonceBytes) bytes of base64url")
        }
        guard let ciphertextText = wire.ct, let sealed = Data(base64URLEncoded: ciphertextText),
            sealed.count >= tagBytes
        else {
            throw .badEnvelope(reason: "ct must be base64url of ciphertext plus GCM tag")
        }
        guard key.count == keyBytes else {
            // Not a v1 Notification Key, so nothing it could ever decrypt.
            throw .decryptFailed
        }

        let plaintext: Data
        do {
            let box = try AES.GCM.SealedBox(
                nonce: AES.GCM.Nonce(data: nonce),
                ciphertext: sealed.dropLast(tagBytes),
                tag: sealed.suffix(tagBytes))
            plaintext = try AES.GCM.open(
                box, using: SymmetricKey(data: key), authenticating: additionalData)
        } catch {
            // Tampered ciphertext, tag, or nonce, or the wrong key: GCM
            // cannot tell these apart and neither can we.
            throw .decryptFailed
        }
        return try validated(plaintext)
    }

    private static func validated(
        _ plaintext: Data
    ) throws(NotificationEnvelopeError) -> NotificationPayload {
        let wire: WirePlaintext
        do {
            wire = try JSONDecoder().decode(WirePlaintext.self, from: plaintext)
        } catch {
            throw .badPayload(reason: "plaintext is not a well-formed JSON object")
        }
        guard let pane = wire.pane, !pane.isEmpty else {
            throw .badPayload(reason: "pane must be a non-empty string")
        }
        guard let kind = wire.kind, !kind.isEmpty else {
            throw .badPayload(reason: "kind must be a non-empty string")
        }
        guard let status = wire.status, !status.isEmpty else {
            throw .badPayload(reason: "status must be a non-empty string")
        }
        guard let rawTimestamp = wire.ts, let timestamp = Int(exactly: rawTimestamp),
            timestamp > 0
        else {
            throw .badPayload(reason: "ts must be a positive unix-seconds integer")
        }
        return NotificationPayload(
            paneID: pane, agentKind: kind, status: AgentStatus(rawValue: status),
            timestamp: Date(timeIntervalSince1970: TimeInterval(timestamp)))
    }

    /// Cleartext JSON wire shape. Every field is optional and numbers are
    /// decoded as Double so that missing keys and wrong types surface as
    /// `badEnvelope` after decoding instead of dying inside JSONDecoder.
    private struct WireEnvelope: Decodable {
        var v: Double?
        var kid: String?
        var n: String?
        var ct: String?
    }

    /// Decrypted JSON wire shape, lenient for the same reason.
    private struct WirePlaintext: Decodable {
        var pane: String?
        var kind: String?
        var status: String?
        var ts: Double?
    }
}

/// Why an envelope did not yield a payload. These map 1:1 to the error
/// identifiers in the shared test vectors. Any of them means the service
/// extension shows the generic fallback banner instead of decrypted content.
enum NotificationEnvelopeError: Error, Sendable, Equatable {
    /// The cleartext framing is broken: not a JSON object, or a missing or
    /// mistyped field.
    case badEnvelope(reason: String)
    /// An envelope from a plugin speaking another contract version.
    case unsupportedVersion(found: Int)
    /// GCM authentication failed: tampered material or the wrong key.
    case decryptFailed
    /// The ciphertext opened but its plaintext violates the payload schema.
    case badPayload(reason: String)

    /// The cross-implementation identifier used by the shared test vectors.
    var wireCode: String {
        switch self {
        case .badEnvelope: "bad_envelope"
        case .unsupportedVersion: "unsupported_version"
        case .decryptFailed: "decrypt_failed"
        case .badPayload: "bad_payload"
        }
    }
}

File: android/core/src/main/java/dev/bybee/heeler/core/crypto/NotificationEnvelope.kt
package dev.bybee.heeler.core.crypto

import java.math.BigDecimal
import java.security.GeneralSecurityException
import java.security.MessageDigest
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec
import kotlinx.serialization.SerializationException
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

/** A validated Agent-status transition carried in an encrypted notification. */
data class NotificationPayload(
    val paneId: String,
    val agentKind: String,
    val status: String,
    val timestamp: Long,
    val project: String? = null,
    val title: String? = null,
)

/** AES-256-GCM notification envelope v1, shared with the Host plugin. */
object NotificationEnvelope {
    const val VERSION = 1
    private const val KEY_BYTES = 32
    private const val NONCE_BYTES = 12
    private const val TAG_BYTES = 16
    private const val KEY_ID_BYTES = 8
    private const val DISPLAY_FIELD_MAX = 256
    private val aad = "HERDR-NOTIFY:$VERSION".toByteArray(Charsets.UTF_8)
    private val json = Json { ignoreUnknownKeys = true }

    /** Derives the unpadded base64url id used to select a Host's Notification Key. */
    fun keyId(key: ByteArray): String {
        require(key.size == KEY_BYTES) { "Notification Key must be $KEY_BYTES bytes" }
        val digest = MessageDigest.getInstance("SHA-256").digest(key)
        return Base64Url.encode(digest.copyOfRange(0, KEY_ID_BYTES))
    }

    /** Returns a structurally valid key id without attempting decryption. */
    fun peekKeyId(envelope: String): String? {
        val wire = try {
            json.parseToJsonElement(envelope) as? JsonObject
        } catch (_: SerializationException) {
            null
        } ?: return null
        val keyId = wire["kid"].stringOrNull() ?: return null
        return keyId.takeIf { Base64Url.decode(it)?.size == KEY_ID_BYTES }
    }

    /**
     * Encrypts a payload into the compact canonical v1 JSON envelope. [nonce] exists so the
     * cross-platform vectors can assert exact output; production callers leave it null.
     */
    fun encrypt(
        payload: NotificationPayload,
        key: ByteArray,
        nonce: ByteArray? = null,
        random: SecureRandom = SecureRandom(),
    ): String {
        require(key.size == KEY_BYTES) { "Notification Key must be $KEY_BYTES bytes" }
        validatePayload(payload)
        val actualNonce = nonce?.copyOf() ?: ByteArray(NONCE_BYTES).also(random::nextBytes)
        require(actualNonce.size == NONCE_BYTES) { "GCM nonce must be $NONCE_BYTES bytes" }

        val plaintext = encodePayload(payload).toByteArray(Charsets.UTF_8)
        val ciphertext = try {
            Cipher.getInstance("AES/GCM/NoPadding").apply {
                init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(TAG_BYTES * 8, actualNonce))
                updateAAD(aad)
            }.doFinal(plaintext)
        } catch (error: GeneralSecurityException) {
            throw IllegalStateException("AES-GCM unavailable", error)
        }
        return buildString {
            append("{\"v\":")
            append(VERSION)
            append(",\"kid\":")
            append(json.encodeToString(keyId(key)))
            append(",\"n\":")
            append(json.encodeToString(Base64Url.encode(actualNonce)))
            append(",\"ct\":")
            append(json.encodeToString(Base64Url.encode(ciphertext)))
            append('}')
        }
    }

    /** Decrypts and validates a v1 envelope with the error taxonomy used by the shared vectors. */
    fun decrypt(envelope: String, key: ByteArray): NotificationPayload {
        val element = try {
            json.parseToJsonElement(envelope)
        } catch (_: SerializationException) {
            throw NotificationEnvelopeError.BadEnvelope
        }
        val wire = element as? JsonObject ?: throw NotificationEnvelopeError.BadEnvelope
        val version = wire["v"].integerOrNull() ?: throw NotificationEnvelopeError.BadEnvelope
        if (version != VERSION.toLong()) throw NotificationEnvelopeError.UnsupportedVersion(version)
        wire["kid"].stringOrNull()
            ?.takeIf { Base64Url.decode(it)?.size == KEY_ID_BYTES }
            ?: throw NotificationEnvelopeError.BadEnvelope
        val nonce = wire["n"].stringOrNull()?.let(Base64Url::decode)
            ?.takeIf { it.size == NONCE_BYTES }
            ?: throw NotificationEnvelopeError.BadEnvelope
        val ciphertext = wire["ct"].stringOrNull()?.let(Base64Url::decode)
            ?.takeIf { it.size >= TAG_BYTES }
            ?: throw NotificationEnvelopeError.BadEnvelope
        if (key.size != KEY_BYTES) throw NotificationEnvelopeError.DecryptFailed

        val plaintext = try {
            Cipher.getInstance("AES/GCM/NoPadding").apply {
                init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(TAG_BYTES * 8, nonce))
                updateAAD(aad)
            }.doFinal(ciphertext)
        } catch (_: GeneralSecurityException) {
            throw NotificationEnvelopeError.DecryptFailed
        }
        return decodePayload(plaintext)
    }

    private fun encodePayload(payload: NotificationPayload): String = buildString {
        append("{\"pane\":")
        append(json.encodeToString(payload.paneId))
        append(",\"kind\":")
        append(json.encodeToString(payload.agentKind))
        append(",\"status\":")
        append(json.encodeToString(payload.status))
        append(",\"ts\":")
        append(payload.timestamp)
        payload.project?.takeIf(String::isNotEmpty)?.let {
            append(",\"project\":")
            append(json.encodeToString(it))
        }
        payload.title?.takeIf(String::isNotEmpty)?.let {
            append(",\"title\":")
            append(json.encodeToString(it))
        }
        append('}')
    }

    private fun decodePayload(plaintext: ByteArray): NotificationPayload {
        val text = Base64Url.decodeUtf8(plaintext) ?: throw NotificationEnvelopeError.BadPayload
        val element = try {
            json.parseToJsonElement(text)
        } catch (_: SerializationException) {
            throw NotificationEnvelopeError.BadPayload
        }
        val wire = element as? JsonObject ?: throw NotificationEnvelopeError.BadPayload
        val paneId = wire["pane"].stringOrNull()?.takeIf(String::isNotEmpty)
            ?: throw NotificationEnvelopeError.BadPayload
        val agentKind = wire["kind"].stringOrNull()?.takeIf(String::isNotEmpty)
            ?: throw NotificationEnvelopeError.BadPayload
        val status = wire["status"].stringOrNull()?.takeIf(String::isNotEmpty)
            ?: throw NotificationEnvelopeError.BadPayload
        val timestamp = wire["ts"].integerOrNull()?.takeIf { it > 0 }
            ?: throw NotificationEnvelopeError.BadPayload
        return NotificationPayload(
            paneId = paneId,
            agentKind = agentKind,
            status = status,
            timestamp = timestamp,
            project = wire.optionalDisplayString("project"),
            title = wire.optionalDisplayString("title"),
        )
    }

    private fun validatePayload(payload: NotificationPayload) {
        if (payload.paneId.isEmpty() || payload.agentKind.isEmpty() || payload.status.isEmpty() || payload.timestamp <= 0) {
            throw NotificationEnvelopeError.BadPayload
        }
        if ((payload.project?.length ?: 0) > DISPLAY_FIELD_MAX || (payload.title?.length ?: 0) > DISPLAY_FIELD_MAX) {
            throw NotificationEnvelopeError.BadPayload
        }
    }

    private fun JsonObject.optionalDisplayString(name: String): String? {
        val value = this[name] ?: return null
        if (value is JsonNull) return null
        val text = value.stringOrNull() ?: throw NotificationEnvelopeError.BadPayload
        if (text.length > DISPLAY_FIELD_MAX) throw NotificationEnvelopeError.BadPayload
        return text.takeIf(String::isNotEmpty)
    }

    private fun JsonElement?.stringOrNull(): String? {
        val primitive = this as? JsonPrimitive ?: return null
        return primitive.content.takeIf { primitive.isString }
    }

    private fun JsonElement?.integerOrNull(): Long? {
        val primitive = this as? JsonPrimitive ?: return null
        if (primitive.isString) return null
        return try {
            BigDecimal(primitive.content).longValueExact()
        } catch (_: NumberFormatException) {
            null
        } catch (_: ArithmeticException) {
            null
        }
    }
}

/** Why an envelope could not produce a notification payload. */
sealed class NotificationEnvelopeError(val wireCode: String) : Exception() {
    data object BadEnvelope : NotificationEnvelopeError("bad_envelope")
    class UnsupportedVersion(val found: Long) : NotificationEnvelopeError("unsupported_version")
    data object DecryptFailed : NotificationEnvelopeError("decrypt_failed")
    data object BadPayload : NotificationEnvelopeError("bad_payload")
}
